### DISCUSSION TITLE
Relationship  between --endpoint and --server-address

### DISCUSSION THREAD

**Discussion by Aptimex** [sandialabs/wiretap#42] [General]
@luker983 

For the `add server` command, the `--server-address` argument is not marked as required. The help message says it "connects to client by default." What exactly does this mean? What happens if you don't specify a value, and try to connect (via `--endpoint`) to an existing server? an existing client? 

Same questions for the same argument in the `add client` command.

**Comment by luker983** [comment]
**`add server` case:**

From the README:
```
If you want to attach a server to another server you must also specify the existing server's API address in your add server command with the --server-address argument;
```

The `server-address` argument is the API address of the server the new server will attach to. If this argument is not supplied, the new server will be "attached" to the client, meaning the client will get a new WG peer in its configuration and the expectation is that `--endpoint` will be the client address (if `--outbound` is not set)

---

**`add client` case**:

The default here is to attach new clients to all of the "relay nodes", these are servers that are one hop from the initial client. If you want to attach a client to any other server, you will need to set `--server-address` to the API address of the target server.

---

In both cases, if you don't specify a server address but *do* set your endpoint to a server incorrectly, those servers won't be properly configured and your new client/server won't work.

**Reply by Aptimex** [reply]
`add server` case:

So `add server -e <client_IP>:<client_port> -r 1.2.3.4/24` will attach the new server to the (first?) existing client, in much the same way the first server and client were attached to each other when the `configure` command was ran? Basically creating a new "relay server"? 

If I got that right we probably need to add some well-defined terms for the special first client and relay servers in the README. 

---

`add client` case:

Does this then assume that the new client will be reachable at the same `--endpoint` by all the relay servers? Similarly, how does (or should) `--outbound` interpret one `--endpoint` value if there are multiple relay servers to connect to?

**Reply by luker983** [reply]
> Basically creating a new "relay server"?

Yup! This is part of the reason servers must be created *before* new clients. Because we don't have a way to propagate new server info to multiple clients.

> Does this then assume that the new client will be reachable at the same --endpoint by all the relay servers? Similarly, how does (or should) --outbound interpret one --endpoint value if there are multiple relay servers to connect to?

Yes, it does assume all relay servers can reach the new client at the same endpoint. Good point about `--outbound`. Currently, `--outbound` has the effect of ignoring the `--endpoint` argument and just copying the existing endpoints for each relay server into the new client's config.
